From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Thu, 14 Aug 2025 19:59:13 +0200
Subject: js/environment: Add implementation for GDesktopAppIcon methods

This is a more complete workaround to cover the GLib bug [1], so that
our extensions are not affected by this.

[1] https://gitlab.gnome.org/GNOME/glib/-/issues/3744
---
 js/ui/environment.js | 28 ++++++++++++----------------
 1 file changed, 12 insertions(+), 16 deletions(-)

diff --git a/js/ui/environment.js b/js/ui/environment.js
index a008a37..9015c00 100644
--- a/js/ui/environment.js
+++ b/js/ui/environment.js
@@ -343,25 +343,21 @@ Object.prototype.toString = function () {
 
 // Work around https://gitlab.gnome.org/GNOME/glib/-/issues/3744
 if (!GioUnix.DesktopAppInfo.prototype.has_key) {
-    GioUnix.DesktopAppInfo.prototype.has_key = function (key) {
-        return GioUnix.DesktopAppInfo.has_key(this, key);
-    };
-
-    GioUnix.DesktopAppInfo.prototype.get_boolean = function (key) {
-        return GioUnix.DesktopAppInfo.get_boolean(this, key);
-    };
+    const reservedNames = ['$gtype', 'prototype', 'length', 'name']
+    Object.getOwnPropertyNames(GioUnix.DesktopAppInfo).forEach(prop => {
+        if (reservedNames.includes(prop))
+            return;
 
-    GioUnix.DesktopAppInfo.prototype.get_categories = function () {
-        return GioUnix.DesktopAppInfo.get_categories(this);
-    };
+        if (prop.startsWith('new'))
+            return;
 
-    GioUnix.DesktopAppInfo.prototype.list_actions = function () {
-        return GioUnix.DesktopAppInfo.list_actions(this);
-    };
+        if (GioUnix.DesktopAppInfo.prototype[prop] !== undefined)
+            return;
 
-    GioUnix.DesktopAppInfo.prototype.get_action_name = function (name) {
-        return GioUnix.DesktopAppInfo.get_action_name(this, name);
-    };
+        GioUnix.DesktopAppInfo.prototype[prop] = function (...args) {
+            return GioUnix.DesktopAppInfo[prop](this, ...args);
+        };
+    });
 }
 
 const slowdownEnv = GLib.getenv('GNOME_SHELL_SLOWDOWN_FACTOR');
